File: golang/pkg/agentbay/context_sync.go
package agentbay

import (
	"encoding/json"
	"fmt"
	"regexp"
)

// UploadStrategy defines the upload strategy for context synchronization
type UploadStrategy string

const (
	// UploadBeforeResourceRelease uploads files before resource release
	UploadBeforeResourceRelease UploadStrategy = "UploadBeforeResourceRelease"
)

// DownloadStrategy defines the download strategy for context synchronization
type DownloadStrategy string

const (
	// DownloadAsync downloads files asynchronously
	DownloadAsync DownloadStrategy = "DownloadAsync"
)

// Lifecycle defines the lifecycle options for recycle policy
type Lifecycle string

const (
	// Lifecycle1Day keeps data for 1 day
	Lifecycle1Day Lifecycle = "Lifecycle_1Day"
	// Lifecycle3Days keeps data for 3 days
	Lifecycle3Days Lifecycle = "Lifecycle_3Days"
	// Lifecycle5Days keeps data for 5 days
	Lifecycle5Days Lifecycle = "Lifecycle_5Days"
	// Lifecycle10Days keeps data for 10 days
	Lifecycle10Days Lifecycle = "Lifecycle_10Days"
	// Lifecycle15Days keeps data for 15 days
	Lifecycle15Days Lifecycle = "Lifecycle_15Days"
	// Lifecycle30Days keeps data for 30 days
	Lifecycle30Days Lifecycle = "Lifecycle_30Days"
	// Lifecycle90Days keeps data for 90 days
	Lifecycle90Days Lifecycle = "Lifecycle_90Days"
	// Lifecycle180Days keeps data for 180 days
	Lifecycle180Days Lifecycle = "Lifecycle_180Days"
	// Lifecycle360Days keeps data for 360 days
	Lifecycle360Days Lifecycle = "Lifecycle_360Days"
	// LifecycleForever keeps data permanently (default)
	LifecycleForever Lifecycle = "Lifecycle_Forever"
)

// UploadPolicy defines the upload policy for context synchronization
type UploadPolicy struct {
	// AutoUpload enables automatic upload
	AutoUpload bool `json:"autoUpload"`
	// UploadStrategy defines the upload strategy
	UploadStrategy UploadStrategy `json:"uploadStrategy"`
}

// NewUploadPolicy creates a new upload policy with default values
func NewUploadPolicy() *UploadPolicy {
	return &UploadPolicy{
		AutoUpload:     true,
		UploadStrategy: UploadBeforeResourceRelease,
	}
}

// DownloadPolicy defines the download policy for context synchronization
type DownloadPolicy struct {
	// AutoDownload enables automatic download
	AutoDownload bool `json:"autoDownload"`
	// DownloadStrategy defines the download strategy
	DownloadStrategy DownloadStrategy `json:"downloadStrategy"`
}

// NewDownloadPolicy creates a new download policy with default values
func NewDownloadPolicy() *DownloadPolicy {
	return &DownloadPolicy{
		AutoDownload:     true,
		DownloadStrategy: DownloadAsync,
	}
}

// DeletePolicy defines the delete policy for context synchronization
type DeletePolicy struct {
	// SyncLocalFile enables synchronization of local file deletions
	SyncLocalFile bool `json:"syncLocalFile"`
}

// NewDeletePolicy creates a new delete policy with default values
func NewDeletePolicy() *DeletePolicy {
	return &DeletePolicy{
		SyncLocalFile: true,
	}
}

// ExtractPolicy defines the extract policy for context synchronization
type ExtractPolicy struct {
	// Extract enables file extraction
	Extract bool `json:"extract"`
	// DeleteSrcFile enables deletion of source file after extraction
	DeleteSrcFile bool `json:"deleteSrcFile"`
	// ExtractToCurrentFolder enables extraction to current folder
	ExtractToCurrentFolder bool `json:"extractToCurrentFolder"`
}

// NewExtractPolicy creates a new extract policy with default values
func NewExtractPolicy() *ExtractPolicy {
	return &ExtractPolicy{
		Extract:                true,
		DeleteSrcFile:          true,
		ExtractToCurrentFolder: false,
	}
}

// RecyclePolicy defines the recycle policy for context synchronization
//
// The RecyclePolicy determines how long context data should be retained and which paths
// are subject to the policy.
//
// Lifecycle field determines the data retention period:
//   - Lifecycle1Day: Keep data for 1 day
//   - Lifecycle3Days: Keep data for 3 days
//   - Lifecycle5Days: Keep data for 5 days
//   - Lifecycle10Days: Keep data for 10 days
//   - Lifecycle15Days: Keep data for 15 days
//   - Lifecycle30Days: Keep data for 30 days
//   - Lifecycle90Days: Keep data for 90 days
//   - Lifecycle180Days: Keep data for 180 days
//   - Lifecycle360Days: Keep data for 360 days
//   - LifecycleForever: Keep data permanently (default)
//
// Paths field specifies which directories or files should be subject to the recycle policy:
//   - Must use exact directory/file paths
//   - Wildcard patterns (* ? [ ]) are NOT supported
//   - Empty string "" means apply to all paths in the context
//   - Multiple paths can be specified as a slice
//   - Default: []string{""} (applies to all paths)
type RecyclePolicy struct {
	// Lifecycle defines how long the context data should be retained
	Lifecycle Lifecycle `json:"lifecycle"`
	// Paths specifies which directories or files should be subject to the recycle policy
	Paths []string `json:"paths"`
}

// NewRecyclePolicy creates a new recycle policy with default values
func NewRecyclePolicy() *RecyclePolicy {
	return &RecyclePolicy{
		Lifecycle: LifecycleForever,
		Paths:     []string{""},
	}
}

// isValidLifecycle checks if the given lifecycle value is valid
func isValidLifecycle(lifecycle Lifecycle) bool {
	switch lifecycle {
	case Lifecycle1Day, Lifecycle3Days, Lifecycle5Days, Lifecycle10Days, Lifecycle15Days,
		Lifecycle30Days, Lifecycle90Days, Lifecycle180Days, Lifecycle360Days, LifecycleForever:
		return true
	default:
		return false
	}
}

// Validate validates the RecyclePolicy configuration
func (rp *RecyclePolicy) Validate() error {
	// Validate Lifecycle value
	if !isValidLifecycle(rp.Lifecycle) {
		return fmt.Errorf(
			"invalid lifecycle value: %s. Valid values are: %s, %s, %s, %s, %s, %s, %s, %s, %s, %s",
			rp.Lifecycle,
			Lifecycle1Day, Lifecycle3Days, Lifecycle5Days, Lifecycle10Days, Lifecycle15Days,
			Lifecycle30Days, Lifecycle90Days, Lifecycle180Days, Lifecycle360Days, LifecycleForever,
		)
	}

	// Validate paths don't contain wildcard patterns
	for _, path := range rp.Paths {
		if path != "" && containsWildcard(path) {
			return fmt.Errorf(
				"wildcard patterns are not supported in recycle policy paths. Got: %s. Please use exact directory paths instead",
				path,
			)
		}
	}
	return nil
}

// WhiteList defines the white list configuration
type WhiteList struct {
	// Path is the path to include in the white list
	Path string `json:"path"`
	// ExcludePaths are the paths to exclude from the white list
	ExcludePaths []string `json:"excludePaths,omitempty"`
}

var wildcardPattern = regexp.MustCompile(`[*?\[\]]`)

func containsWildcard(path string) bool {
	return wildcardPattern.MatchString(path)
}

func (wl *WhiteList) Validate() error {
	if containsWildcard(wl.Path) {
		return fmt.Errorf(
			"wildcard patterns are not supported in path. Got: %s. Please use exact directory paths instead",
			wl.Path,
		)
	}

	for _, excludePath := range wl.ExcludePaths {
		if containsWildcard(excludePath) {
			return fmt.Errorf(
				"wildcard patterns are not supported in exclude_paths. Got: %s. Please use exact directory paths instead",
				excludePath,
			)
		}
	}

	return nil
}

// BWList defines the black and white list configuration
type BWList struct {
	// WhiteLists defines the white lists
	WhiteLists []*WhiteList `json:"whiteLists,omitempty"`
}

// SyncPolicy defines the synchronization policy
type SyncPolicy struct {
	// UploadPolicy defines the upload policy
	UploadPolicy *UploadPolicy `json:"uploadPolicy,omitempty"`
	// DownloadPolicy defines the download policy
	DownloadPolicy *DownloadPolicy `json:"downloadPolicy,omitempty"`
	// DeletePolicy defines the delete policy
	DeletePolicy *DeletePolicy `json:"deletePolicy,omitempty"`
	// ExtractPolicy defines the extract policy
	ExtractPolicy *ExtractPolicy `json:"extractPolicy,omitempty"`
	// RecyclePolicy defines the recycle policy
	RecyclePolicy *RecyclePolicy `json:"recyclePolicy,omitempty"`
	// BWList defines the black and white list
	BWList *BWList `json:"bwList,omitempty"`
}

// NewSyncPolicy creates a new sync policy with default values
func NewSyncPolicy() *SyncPolicy {
	return &SyncPolicy{
		UploadPolicy:   NewUploadPolicy(),
		DownloadPolicy: NewDownloadPolicy(),
		DeletePolicy:   NewDeletePolicy(),
		ExtractPolicy:  NewExtractPolicy(),
		RecyclePolicy:  NewRecyclePolicy(),
		BWList: &BWList{
			WhiteLists: []*WhiteList{
				{
					Path:         "",
					ExcludePaths: []string{},
				},
			},
		},
	}
}

// ensureDefaults ensures all policy fields have default values if not provided
func (sp *SyncPolicy) ensureDefaults() {
	if sp.UploadPolicy == nil {
		sp.UploadPolicy = NewUploadPolicy()
	}
	if sp.DownloadPolicy == nil {
		sp.DownloadPolicy = NewDownloadPolicy()
	}
	if sp.DeletePolicy == nil {
		sp.DeletePolicy = NewDeletePolicy()
	}
	if sp.ExtractPolicy == nil {
		sp.ExtractPolicy = NewExtractPolicy()
	}
	if sp.RecyclePolicy == nil {
		sp.RecyclePolicy = NewRecyclePolicy()
	}
	if sp.BWList == nil {
		sp.BWList = &BWList{
			WhiteLists: []*WhiteList{
				{
					Path:         "",
					ExcludePaths: []string{},
				},
			},
		}
	}
}

// MarshalJSON ensures all fields have default values before marshaling
func (sp *SyncPolicy) MarshalJSON() ([]byte, error) {
	sp.ensureDefaults()

	// Create a temporary struct to avoid infinite recursion
	type SyncPolicyAlias SyncPolicy
	return json.Marshal((*SyncPolicyAlias)(sp))
}

// ContextSync defines the context synchronization configuration
type ContextSync struct {
	// ContextID is the ID of the context to synchronize
	ContextID string `json:"contextId"`
	// Path is the path where the context should be mounted
	Path string `json:"path"`
	// Policy defines the synchronization policy
	Policy *SyncPolicy `json:"policy,omitempty"`
}

func validateSyncPolicy(policy *SyncPolicy) error {
	if policy == nil {
		return nil
	}

	// Validate RecyclePolicy paths
	if policy.RecyclePolicy != nil {
		if err := policy.RecyclePolicy.Validate(); err != nil {
			return err
		}
	}

	// Validate BWList whitelist paths
	if policy.BWList != nil && policy.BWList.WhiteLists != nil {
		for _, whitelist := range policy.BWList.WhiteLists {
			if err := whitelist.Validate(); err != nil {
				return err
			}
		}
	}

	return nil
}

// NewContextSync creates a new context sync configuration
func NewContextSync(contextID, path string, policy *SyncPolicy) (*ContextSync, error) {
	if err := validateSyncPolicy(policy); err != nil {
		return nil, err
	}

	return &ContextSync{
		ContextID: contextID,
		Path:      path,
		Policy:    policy,
	}, nil
}

// WithPolicy sets the policy and returns the context sync for chaining
func (cs *ContextSync) WithPolicy(policy *SyncPolicy) (*ContextSync, error) {
	if err := validateSyncPolicy(policy); err != nil {
		return nil, err
	}
	cs.Policy = policy
	return cs, nil
}

File: python/agentbay/context_sync.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Optional
import re


class UploadStrategy(Enum):
    """Upload strategy for context synchronization"""

    UPLOAD_BEFORE_RESOURCE_RELEASE = "UploadBeforeResourceRelease"


class DownloadStrategy(Enum):
    """Download strategy for context synchronization"""

    DOWNLOAD_ASYNC = "DownloadAsync"

class UploadMode(Enum):
    """Upload mode for context synchronization"""
    
    FILE = "File"
    ARCHIVE = "Archive"


class Lifecycle(Enum):
    """Lifecycle options for recycle policy"""
    
    LIFECYCLE_1DAY = "Lifecycle_1Day"
    LIFECYCLE_3DAYS = "Lifecycle_3Days"
    LIFECYCLE_5DAYS = "Lifecycle_5Days"
    LIFECYCLE_10DAYS = "Lifecycle_10Days"
    LIFECYCLE_15DAYS = "Lifecycle_15Days"
    LIFECYCLE_30DAYS = "Lifecycle_30Days"
    LIFECYCLE_90DAYS = "Lifecycle_90Days"
    LIFECYCLE_180DAYS = "Lifecycle_180Days"
    LIFECYCLE_360DAYS = "Lifecycle_360Days"
    LIFECYCLE_FOREVER = "Lifecycle_Forever"


@dataclass
class UploadPolicy:
    """
    Defines the upload policy for context synchronization

    Attributes:
        auto_upload: Enables automatic upload
        upload_strategy: Defines the upload strategy
        upload_mode: Defines the upload mode (UploadMode.FILE or UploadMode.ARCHIVE)
    """

    auto_upload: bool = True
    upload_strategy: UploadStrategy = UploadStrategy.UPLOAD_BEFORE_RESOURCE_RELEASE
    upload_mode: UploadMode = UploadMode.FILE

    def __post_init__(self):
        """Validate upload_mode value"""
        if not isinstance(self.upload_mode, UploadMode):
            valid_values = [e.value for e in UploadMode]
            raise ValueError(
                f"Invalid upload_mode value: {self.upload_mode}. "
                f"Valid values are: {', '.join(valid_values)}"
            )

    @classmethod
    def default(cls):
        """Creates a new upload policy with default values"""
        return cls()

    def __dict__(self):
        return {
            "autoUpload": self.auto_upload,
            "uploadStrategy": (
                self.upload_strategy.value if self.upload_strategy else None
            ),
            "uploadMode": self.upload_mode.value if self.upload_mode else None,
        }


@dataclass
class DownloadPolicy:
    """
    Defines the download policy for context synchronization

    Attributes:
        auto_download: Enables automatic download
        download_strategy: Defines the download strategy
    """

    auto_download: bool = True
    download_strategy: DownloadStrategy = DownloadStrategy.DOWNLOAD_ASYNC

    @classmethod
    def default(cls):
        """Creates a new download policy with default values"""
        return cls()

    def __dict__(self):
        return {
            "autoDownload": self.auto_download,
            "downloadStrategy": (
                self.download_strategy.value if self.download_strategy else None
            ),
        }


@dataclass
class DeletePolicy:
    """
    Defines the delete policy for context synchronization

    Attributes:
        sync_local_file: Enables synchronization of local file deletions
    """

    sync_local_file: bool = True

    @classmethod
    def default(cls):
        """Creates a new delete policy with default values"""
        return cls()

    def __dict__(self):
        return {"syncLocalFile": self.sync_local_file}


@dataclass
class ExtractPolicy:
    """
    Defines the extract policy for context synchronization

    Attributes:
        extract: Enables file extraction
        delete_src_file: Enables deletion of source file after extraction
    """

    extract: bool = True
    delete_src_file: bool = True
    extract_current_folder: bool = False

    @classmethod
    def default(cls):
        """Creates a new extract policy with default values"""
        return cls()

    def __dict__(self):
        return {"extract": self.extract, "deleteSrcFile": self.delete_src_file, "extractToCurrentFolder": self.extract_current_folder}


@dataclass
class RecyclePolicy:
    """
    Defines the recycle policy for context synchronization
    
    Attributes:
        lifecycle: Defines how long the context data should be retained
            Available options:
            - LIFECYCLE_1DAY: Keep data for 1 day
            - LIFECYCLE_3DAYS: Keep data for 3 days
            - LIFECYCLE_5DAYS: Keep data for 5 days
            - LIFECYCLE_10DAYS: Keep data for 10 days
            - LIFECYCLE_15DAYS: Keep data for 15 days
            - LIFECYCLE_30DAYS: Keep data for 30 days
            - LIFECYCLE_90DAYS: Keep data for 90 days
            - LIFECYCLE_180DAYS: Keep data for 180 days
            - LIFECYCLE_360DAYS: Keep data for 360 days
            - LIFECYCLE_FOREVER: Keep data permanently (default)
        paths: Specifies which directories or files should be subject to the recycle policy
            Rules:
            - Must use exact directory/file paths
            - Wildcard patterns (* ? [ ]) are NOT supported
            - Empty string "" means apply to all paths in the context
            - Multiple paths can be specified as a list
            Default: [""] (applies to all paths)
    """
    
    lifecycle: Lifecycle = Lifecycle.LIFECYCLE_FOREVER
    paths: List[str] = field(default_factory=lambda: [""])

    def __post_init__(self):
        """Validate lifecycle and paths configuration"""
        # Validate lifecycle value
        if not isinstance(self.lifecycle, Lifecycle):
            valid_values = [e.value for e in Lifecycle]
            raise ValueError(
                f"Invalid lifecycle value: {self.lifecycle}. "
                f"Valid values are: {', '.join(valid_values)}"
            )
        
        # Validate that paths don't contain wildcard patterns
        for path in self.paths:
            if path and path.strip() != "" and self._contains_wildcard(path):
                raise ValueError(
                    f"Wildcard patterns are not supported in recycle policy paths. Got: {path}. "
                    "Please use exact directory paths instead."
                )

    @staticmethod
    def _contains_wildcard(path: str) -> bool:
        """Check if path contains wildcard characters"""
        return bool(re.search(r'[*?\[\]]', path))

    @classmethod
    def default(cls):
        """Creates a new recycle policy with default values"""
        return cls()

    def __dict__(self):
        return {
            "lifecycle": self.lifecycle.value if self.lifecycle else None,
            "paths": self.paths
        }


@dataclass
class WhiteList:
    """
    Defines the white list configuration

    Attributes:
        path: Path to include in the white list
        exclude_paths: Paths to exclude from the white list
    """

    path: str = ""
    exclude_paths: List[str] = field(default_factory=list)

    def __post_init__(self):
        """Validate that paths don't contain wildcard patterns"""
        if self._contains_wildcard(self.path):
            raise ValueError(
                f"Wildcard patterns are not supported in path. Got: {self.path}. "
                "Please use exact directory paths instead."
            )
        for exclude_path in self.exclude_paths:
            if self._contains_wildcard(exclude_path):
                raise ValueError(
                    f"Wildcard patterns are not supported in exclude_paths. Got: {exclude_path}. "
                    "Please use exact directory paths instead."
                )

    @staticmethod
    def _contains_wildcard(path: str) -> bool:
        """Check if path contains wildcard characters"""
        return bool(re.search(r'[*?\[\]]', path))

    def __dict__(self):
        return {"path": self.path, "excludePaths": self.exclude_paths}


@dataclass
class BWList:
    """
    Defines the black and white list configuration

    Attributes:
        white_lists: Defines the white lists
    """

    white_lists: List[WhiteList] = field(default_factory=list)

    def __dict__(self):
        return {
            "whiteLists": (
                [wl.__dict__() for wl in self.white_lists] if self.white_lists else []
            )
        }


@dataclass
class MappingPolicy:
    """
    Defines the mapping policy for cross-platform context synchronization

    Attributes:
        path: The original path from a different OS that should be mapped to the current context path
    """

    path: str = ""

    @classmethod
    def default(cls):
        """Creates a new mapping policy with default values"""
        return cls()

    def __dict__(self):
        return {"path": self.path}


@dataclass
class SyncPolicy:
    """
    Defines the synchronization policy

    Attributes:
        upload_policy: Defines the upload policy
        download_policy: Defines the download policy
        delete_policy: Defines the delete policy
        extract_policy: Defines the extract policy
        recycle_policy: Defines the recycle policy
        bw_list: Defines the black and white list
        mapping_policy: Defines the mapping policy for cross-platform context synchronization
    """

    upload_policy: Optional[UploadPolicy] = None
    download_policy: Optional[DownloadPolicy] = None
    delete_policy: Optional[DeletePolicy] = None
    extract_policy: Optional[ExtractPolicy] = None
    recycle_policy: Optional[RecyclePolicy] = None
    bw_list: Optional[BWList] = None
    mapping_policy: Optional[MappingPolicy] = None

    def __post_init__(self):
        """Post-initialization to ensure all policies have default values if not provided"""
        if self.upload_policy is None:
            self.upload_policy = UploadPolicy.default()
        if self.download_policy is None:
            self.download_policy = DownloadPolicy.default()
        if self.delete_policy is None:
            self.delete_policy = DeletePolicy.default()
        if self.extract_policy is None:
            self.extract_policy = ExtractPolicy.default()
        if self.recycle_policy is None:
            self.recycle_policy = RecyclePolicy.default()
        if self.bw_list is None:
            self.bw_list = BWList(white_lists=[WhiteList(path="", exclude_paths=[])])

    @classmethod
    def default(cls):
        """Creates a new sync policy with default values"""
        return cls(
            upload_policy=UploadPolicy.default(),
            download_policy=DownloadPolicy.default(),
            delete_policy=DeletePolicy.default(),
            extract_policy=ExtractPolicy.default(),
            recycle_policy=RecyclePolicy.default(),
            bw_list=BWList(white_lists=[WhiteList(path="", exclude_paths=[])]),
        )

    def __dict__(self):
        result = {}
        if self.upload_policy:
            result["uploadPolicy"] = self.upload_policy.__dict__()
        if self.download_policy:
            result["downloadPolicy"] = self.download_policy.__dict__()
        if self.delete_policy:
            result["deletePolicy"] = self.delete_policy.__dict__()
        if self.extract_policy:
            result["extractPolicy"] = self.extract_policy.__dict__()
        if self.recycle_policy:
            result["recyclePolicy"] = self.recycle_policy.__dict__()
        if self.bw_list:
            result["bwList"] = self.bw_list.__dict__()
        if self.mapping_policy:
            result["mappingPolicy"] = self.mapping_policy.__dict__()
        return result


@dataclass
class ContextSync:
    """
    Defines the context synchronization configuration

    Attributes:
        context_id: ID of the context to synchronize
        path: Path where the context should be mounted
        policy: Defines the synchronization policy
    """

    context_id: str
    path: str
    policy: Optional[SyncPolicy] = None

    @classmethod
    def new(cls, context_id: str, path: str, policy: Optional[SyncPolicy] = None):
        """Creates a new context sync configuration"""
        return cls(context_id=context_id, path=path, policy=policy)

    def with_policy(self, policy: SyncPolicy):
        """Sets the policy"""
        self.policy = policy
        return self